### ISSUE TITLE
False positive: The value of the associated type `Output` (from trait `FnOnce`) must be specified [E0191]

### ISSUE THREAD

**Issue by amoskvin** [intellij-rust/intellij-rust#10633]
## Environment

* **IntelliJ Rust plugin version:** 0.4.197.5401-231
* **Rust toolchain version:** 1.70.0 (ec8a8a0ca 2023-04-25) x86_64-unknown-linux-gnu
* **IDE name and version:** CLion 2023.1.4 (CL-231.9161.40)
* **Operating system:** Linux
* **Macro expansion:** enabled

## Problem description
The code below compiles, but the plugin gives a false E0191

## Steps to reproduce
```rust
trait Callback: FnMut(i32) {}

impl<F> Callback for F where F: FnMut(i32) {}

fn func(mut callback: Box<dyn Callback>) {
    //                        ^^^^^^^^
    (*callback)(42)
}

enum Enum {
    Unit(Box<dyn Callback>),
    //           ^^^^^^^^
}

fn main() {
    func(Box::new(|x| println!("The answer is {x}")));
}
```

**Comment by Maksim20023** [comment]
Hi! Thank you for reporting the issue and detailed description. According to this, I reproduced the problem. We’ll investigate the reasons for such behavior and add a reply when we will have additional information about it. If you have any further suggestions, please feel free to reach out.
